orquesta/tests/unit/conducting/test_workflow_conductor_rerun.py

```diff
@@ -0,0 +1,599 @@
+        self.forward_task_statuses(conductor, 'task2', [statuses.RUNNING])
+
+        self.forward_task_statuses(conductor, 'task1', [statuses.SUCCEEDED])
+        self.forward_task_statuses(conductor, 'task2', [statuses.SUCCEEDED])
+
+        self.forward_task_statuses(conductor, 'task3', [statuses.RUNNING])
+        self.forward_task_statuses(conductor, 'task3', [statuses.FAILED])
+
+        # Check the list of terminal tasks.
+        term_tasks = conductor.workflow_state.get_terminal_tasks()
+        actual_term_tasks = [t['id'] for t in term_tasks]
+        expected_term_tasks = ['task3']
+        self.assertListEqual(actual_term_tasks, expected_term_tasks)
+        self.assertEqual(conductor.get_workflow_status(), statuses.FAILED)
+        conductor.reset_workflow_output()
+
+        # 3. Start rerun workflow
+        rerun_tasks = ['task3']
+        number_tasks = 3
+        self.assert_workflow_rerun_request(conductor, rerun_tasks, number_tasks)
+
+        # Conduct task3 and check merged context.
+        self.forward_task_statuses(conductor, 'task3', [statuses.RUNNING])
+        self.forward_task_statuses(conductor, 'task3', [statuses.RUNNING, statuses.SUCCEEDED])
+
+        expected_task_ctx = {'var1': 'xyz', 'var2': 123}
+        expected_task_init_ctx = expected_task_ctx
+        self.assertDictEqual(conductor.get_task_initial_context('task3', 0), expected_task_init_ctx)
+        expected_task_ctx = dict_util.merge_dicts(expected_task_ctx, {'var3': True})
+        expected_txsn_ctx = {'task4__t0': expected_task_ctx}
+        self.assertDictEqual(conductor.get_task_transition_contexts('task3', 0), expected_txsn_ctx)
+        self.assert_next_task(conductor, 'task4', expected_task_ctx)
+
+        # Conduct task4 and check final workflow status.
+        self.forward_task_statuses(conductor, 'task4', [statuses.RUNNING, statuses.SUCCEEDED])
+        self.assert_next_task(conductor, has_next_task=False)
+
+        # Check the list of terminal tasks.
+        term_tasks = conductor.workflow_state.get_terminal_tasks()
+        actual_term_tasks = [t['id'] for t in term_tasks]
+        expected_term_tasks = ['task3', 'task4']
+        self.assertListEqual(actual_term_tasks, expected_term_tasks)
+
+        # Check workflow status and context.
+        expected_term_ctx = copy.deepcopy(expected_task_ctx)
+        self.assertDictEqual(conductor.get_workflow_terminal_context(), expected_term_ctx)
+        self.assertEqual(conductor.get_workflow_status(), statuses.SUCCEEDED)
+
+    def test_self_looping_rerun(self):
+        wf_def = """
+        version: 1.0
+
+        description: A basic workflow with cycle.
+
+        vars:
+          - loop: True
+
+        tasks:
+          task1:
+            action: core.noop
+            next:
+              - do: task2
+          task2:
+            action: core.noop
+            next:
+              - when: <% ctx(loop) = true %>
```

Comment:
Since we control the status of the execution, you just need to check `when: <% failed() %>` then `do: task2`.  We don't need to check loop variable and publish loop to False.

Reply:
I see what you mean now. I was thinking about something else. But it doesn't matter, this test is testing what it's suppose to. Let's leave this use case as is.